### ISSUE TITLE
What happens when laptop lid is closed / phone display is turned off?

### ISSUE THREAD

**Issue by jan-ivar** [w3c/mediacapture-main#670]
Seperate from https://github.com/w3c/mediacapture-main/issues/668 because, as mentioned in https://github.com/w3c/webrtc-pc/issues/2489#issuecomment-602879082:
> 1. Closing a laptop lid is _**not**_ the same as freezing. It may _trigger_ it, but was observed not to.
> 2. It's a physical act that (sans external display) seizes display, audio output, mic & cam capture.

This physical act seems similar (but not identical) to pressing the physical "display off" button on most smartphones and tablets, or closing a flip phone.

To complicate matters, there appear to be several device exceptions:
1. A laptop with external monitor attached may not seize media i/o.
2. A flip phone with back-display may have its back camera work as front camera when closed.
3. Some phone apps (like YouTube now) still play audio while screen is off (with or wo/headset)

But one commonality seems to be that most devices have some way to shut media playback up temporarily, and resume later. Should we piggyback media capture on that?

Secondly, we already have the `mute` and `unmute` user agent events for notifying JS. But from my observations, browsers aren't firing them when I close and reopen the laptop lid, even though capture is interrupted. Do we need to say anything more, or simply file bugs on implementations?

Thirdly, there may be common privacy issues around this behavior. E.g. Users may be surprised
 1. if microphone keeps recording when they switch display off phone/tablet to put in pocket.
 2. if laptop microphone keeps recording when they close the laptop lid.
 3. USB camera keeps recording when they close the laptop lid.
 4. If camera and mic (and screen-sharing?) resume during laptop lock screen on reopen
 5. If camera and mic don't shut off when audible playback shuts off.
 6. If camera and mic resumes before audible playback resumes.

Do we want to try to say anything around that?

**Comment by rshpount** [comment]
The physical "display off" button on smartphones does not seem to work correctly either. If you are on the call via WebRTC in Chrome on Android, switching display off stops microphone and playback a short time after the screen is turned of. At the same time, if you are on the regular phone call, switching display off does not disconnect the call or mutes the microphone.  This is one of the common customer complaints from people using our WebRTC application from smartphones.

**Comment by jan-ivar** [comment]
Good point that user expectations for "display off" on phones/tablets may differ from laptop lid.

**Comment by guest271314** [comment]
> Do we need to say anything more, or simply file bugs on implementations?

What is the bug? 

Why cannot a user decide to close their laptop lid and transmit and record audio from `getUserMedia()`?

> 1. if microphone keeps recording when they switch display off phone/tablet to put in pocket.
if laptop microphone keeps recording when they close the laptop lid.
> 2. if microphone keeps recording when they switch display off phone/tablet to put in pocket.
if laptop microphone keeps recording when they close the laptop lid.

There are use cases for microphone recording when laptop lid is closed and display is "off". These settings can be configured at OS level. Why is `getUserMedia()` attempting to define the behaviour of a laptop lid being closed?

**Comment by guest271314** [comment]
At Nightly 76 and Chromium 82 when laptop lid is closed after `getUserMedia()` is executed - and the setting is specifically "When laptop lid is closed: Suspend" - the setting ha no effect. Am able to record audio while laptop lid is closed without the Power Manager setting affecting `MediaStream` or `MediaStreamTrack`, as expected. 

The "Suspend" option has the same effect as the "Lock Screen" option at the OS where tested.

That is is just using a Power Manager GUI, without attempting to navigate around whatever result is being considered here re Media Capture and Streams somehow attempting to access the device settings and state.

`MediaStreamTrack` does not fire `mute` or `unmute` events.
![Screenshot_2020-03-27_02-32-23](https://user-images.githubusercontent.com/4174848/77715865-ad54ff00-6fd4-11ea-9f7b-ab2307bc8af1.png)

**Comment by guest271314** [comment]
@jan-ivar 

>  But from my observations, browsers aren't firing them when I close and reopen the laptop lid, even though capture is interrupted.

What evidence do you have that capture is interrupted when laptop lid is closed?

Is  not 

> What happens when laptop lid is closed / phone display is turned off?

per the specification "out of control for the application"?

> Muted is out of control for the application

Similar to configuring power ("suspend"; "hibernate") at laptop/desktop cannot a user who has specific hardware and OS requirements re Media Capture and Streams (or any other hardware setting or OS source code) root their "smart" phone and set whatever configuration they want?

**Comment by jan-ivar** [comment]
> What is the bug?

I should have mentioned I'm on macOS, where browsers today [[1](https://bugzilla.mozilla.org/show_bug.cgi?id=1622073)] [[2](https://bugs.chromium.org/p/chromium/issues/detail?id=1061204)] by default temporarily pause mic and camera recording (when I close my MBP lid) without firing `mute`, while JS remains running. This violates the existing [spec](https://w3c.github.io/mediacapture-main/getusermedia.html#track-muted): *"Whenever the User Agent initiates such a* [physical] *change, it MUST queue a task, ... to... Fire a simple event named* [mute] *on track."*

> What evidence do you have that capture is interrupted when laptop lid is closed?

See above. I'm not claiming this always happens on all systems or even all configurations.

> Why is getUserMedia() attempting to define the behaviour of a laptop lid being closed?

I'm not suggesting that in the [three things](https://github.com/w3c/mediacapture-main/issues/670#issue-586571136) I propose.

1. Shut off media capture whenever media playback shuts off for physical reasons
2. Here I specifically said _"browsers aren't firing them ... even though capture is interrupted."_
3. Mention privacy issues/guidelines for vendors to avoid users being "surprised" recording is on.

**Comment by guest271314** [comment]
Am not sure what "MBP" is.

The OS should not pause camera and microphone recording simply because the laptop lid is closed. 

The user could still want to transmit and receive data while the laptop lid is closed, for a variety of use cases, and this specification should not include language arbitrarily stop transmitting data just because the device lid is closed. 

A user could be recording multiple hours of video and audio and does not need to view the screen or listen to audio output - or allow a passerby to view or hear the process - to perform those tasks. 

> "Whenever the User Agent initiates such a [physical] change, it MUST queue a task, ... to... Fire a simple event named [mute] on track."

The User-Agent is the browser, not the machine itself or the OS. That appears to be an OS bug, not a problem with Media Capture and Streams.

Given the human MUST physically close the lid on the laptop, the user can physically execute `MediaStreamTrack.stop()` or `MediaStreamTrack.enabled = false` at their keyboard _before_ closing the lid on the device.

**Comment by jan-ivar** [comment]
MBP = Macbook Pro.

> The OS should not pause camera and microphone recording simply because the laptop lid is closed.

That's out of scope of this spec. Take it up with the manufacturer.

> The User-Agent is the browser, not the machine itself or the OS. 

Here's the [full text](https://w3c.github.io/mediacapture-main/getusermedia.html#track-muted): *"There can be several reasons for a MediaStreamTrack to be muted: the user pushing a **physical mute button** on the microphone, the **user toggling a control in the operating system**, the user clicking a mute button in the browser chrome, the User Agent (**on behalf of the user**) mutes, etc.*

The goal is to give a consistent model to the JS application. If frames stop coming in on the track while the track is unmuted, that's confusing. It's better to give JS a reason for this by notifying it. Since it can already observe that frames stop coming in when input is paused, there's no leak of information here.

This does not mean the user closed the lid.

If you're concerned about laptop lid detection, I recommend looking at https://github.com/w3c/webrtc-pc/issues/2489 and https://bugs.chromium.org/p/chromium/issues/detail?id=1060547.

**Comment by guest271314** [comment]
Is "MBP" "Mac Book Pro"? 

Some uses cases of continued transmission of `MediaStreamTrack` data while laptop lid is closed are:

- Creating an audio book. There is no need to view a screen to read a physical book that is recorded  and then transcribed to text with other API's, e.g., speech recognition technologies.
- Voice commands. Note, the voice does not necessarily need to be a live human voice. The laptop lid does not need to be open to provide the voice command to launch a headless version of a browser, navigate to a page, execute `getDisplayMedia()` to grab a still image or record the screen. This can be done multiple times to create a collage of media from fragments which are then merged into one media stream that is edited and exported - again using `MediaStream`.
- Working on a project that is, for example, an exclusive premiere of media. The developer might not want anyone else to be able to view their work in progress, they close the lid on their device for a few moments. Then re-open the lid on the device. All work should not be lost because a specification wants to stop media capture  when the lid is closed. 

> Mention privacy issues/guidelines for vendors to avoid users being "surprised" recording is on.

Users should not be surprised recording is on. They turned it on. They navigated to the site and were prompted to capture their media devices, which they granted permission for. Nor should users rely on automation alone to turn recording off. They did not physically turn off recording, instead they bypassed that critical step to close the machine lid. The simplest approach, since physical  activity by the human is necessary to both type `MediaStreamTrack.stop()` or `MediaStreamTrack.disabled` or to click the relevant UI the implementer provides, is to not omit clicking to revoke permissions _before_ closing the lid on the device. 

The proposed presumptive solution to just stop the stream has the initial theme of not trusting the user to perform their own due diligence, and therefore since the specification authors do not trust the users to handle their own streams - perhaps from a specification author having the occasion not handling their own media stream themselves in a certain case - will now attempt to place the entirety of the world in the category of their case where they expected the stream to stop when _they_ individually close their device lid. 

A rational solution to the case of a user expecting a media stream to stop when the lid is closed is apparently to use a "MBP" machine and *pple OS. 

Even then, since the specification author would, presumably,  consider the opposite of their experiences and use cases as well as their own experiences and use cases, yet another solution is to advise (include a **_Note_**) the user that the media stream MAY or MAY NOT stop capture when lid is closed, then the user is aware and can take steps to configure their power manager at the OS, to either continue capture or stop capture when lid is closed, rather than to have only one option of stopping the stream, which  appears to be  a  would-be  safeguard against the user making  a mistake,  or omitting revoking permissions previously  granted before closing the  lid, which does not make  sense if they truly intend to affirmatively stop the stream -  they  need to do that before closing the lid,   which is the logical steps of the  procedure if that is in fact what they intend to occur: 

1. Do not want media capture to occur when lid is closed.
2. Revoke permissions.
3. Close lid.

Skipping 2. is not logical when the user is  aware of the stream ongoing.

For the case of users who are using a "smart" phone to capture media, well, that entire landscape is rife with issues that this specification  cannot remedy. In essence what is being proposed is the elimination of "butt-dials" or "pocket calls", or in this case, ongoing "pocket media stream captures" where the actual solution for that dilemma is to not walk around with  a "smart" phone in your pocket all of the time. It actually is possible to simply set the device down, and not have a "smart" phone leashed to one's person at all times, while performing all sorts of human activities and  simultaneously expecting the device to not have a button or key pressed unintentionally.

In summary, users  are responsible for their media capture  -  not the specification.

**Comment by guest271314** [comment]
> That's out of scope of this spec. Take it up with the manufacturer.

Agree. That is what you should do. However, in this case the device appears to be doing what you expect it to do. Thus, you are still using the device instead of a device that does not do that. That is a choice.

*nix does not do that. If it did would start a shell and command the machine to not do that.

> Here's the full text: "There can be several reasons for a MediaStreamTrack to be muted: the user pushing a physical mute button on the microphone, the user toggling a control in the operating system, the user clicking a mute button in the browser chrome, the User Agent (on behalf of the user) mutes, etc.

None of that language articulates that the stream should stop because a laptop lid is closed  or a user put their "smart" phone in their pocket and continued performing activities knowing full well a button could be pressed while the device is in their pocket.

> If you're concerned about laptop lid detection, I recommend looking at w3c/webrtc-pc#2489 and https://bugs.chromium.org/p/chromium/issues/detail?id=1060547.

No, am not concerned about that case, nor the wiles of "smart" phone functionality. Am stating this specification should not be concerned with those matters either. Pick a device that does what you want it to do. As you apparently have. That does not mean just because you expect a stream to stop when the laptop lid is closed, or whatever device you happen to be used mutes a track for some reason, others expect the same output.

**Comment by guest271314** [comment]
If the concern is truly "privacy", the solution is to not use a "smart" phone _at all_. As those devices are capable of (in general, _are_) being tracked all of the time, accessed from near or afar without notice from any individual or entity with a cursory knowledge of the architecture and OS, whether media capture is involved or not.

**Comment by guest271314** [comment]
There are various use cases for a `MediaStream` not being `muted` or `ended` when a laptop lid is closed. 

For an individual who might be vision impaired, they do not need a laptop lid open at all after executing `getUserMeda()`.

They could execute `getUserMedia({audio: true})`, close the lid on the laptop, and having programs running that detect non-silence, for example, using `AudioContext`, for whatever reason, user says:

"News headlines"

which is detected by `AudioContext.AnalyzerNode`. "News headlines" are `fetch()`ed, text parsed, `MediaStreamTrack.enabled` or similar code is used to "pause" audio stream input so that when headlines are output to speakers the program does not capture that particular output, the result of N `fetch()` calls are passed to `window.speechSynthesis.speak()` which outputs parsed text as audio to speakers, when complete, `MediaStreamTrack.enabled` is set to `true` again, to reset program to listen for next voice command. None of that procedure requires the laptop lid to be open, and the user certainly not expect for the application to be biased towards only users who are, for example, are using `MediaStreamTrack` to be communicating with visual and auditory constraints set. 

The user could also be vision impaired and an artist. They might even have a website with digital renditions of their work, with HTML markup configured for them to use a voice command to get a specific work, capture the still image, insert that image into a `MediaStream` and communicate their work to other sites, or users via `RTCPeerConnection`, et al. The laptop lid does not need to be open to perform any of those tasks.

**Comment by guest271314** [comment]
Another use case is are laptops with external camera connected streaming video to N peers. There is no need for the laptop lids to be open when used solely to stream video for scientific, security, media creation (e.g., time lapse) or other purposes. 

One simple solution to avoid media capture when lap top lid is closed or phone display is turned off is to configure the device to turn off networking when the laptop lid is closed.

That is not a solution for an individual that walks around with their phone in their pocket all of the time. The same button which is intentionally set to display off by the user could be unintentionally pressed to display on in their pocket. That is a human condition.

**Comment by alvestrand** [comment]
Likely recommendation: "Nothing".
To be addressed in a best practice statement together with #668.
